### DISCUSSION TITLE
If workerA fails to steal-task and puts itself in the sleeper-queue with is_last_searcher=true, then during the subsequent check for task-existence, other threads may wake up workerA. Isn't this contradictory?

### DISCUSSION THREAD

**Discussion by sxstd001** [tokio-rs/tokio#6072] [Q&A]
**I don't understand:** 
If **workerA** fails to steal-task and puts itself in the sleeper-queue with is_last_searcher=true, then during the subsequent check for task-existence, other threads may wake up workerA. Isn't this contradictory? WorkerA ignores these wake-ups and directly calls self.park_timeout(core, None). This seems problematic!

**Comment by Darksonn** [comment]
It calls `notify_if_work_pending` before going to sleep. If that call finds any work, then `park` will be notified, preventing the thread from going to sleep.

This means that if new work is spawned concurrently with going to sleep, then one of these cases happen:

 * Case 1. The new work is spawned before the runtime calls `transition_worker_to_parked`. In this case, the worker will notice the work in `notify_if_work_pending`, preventing it from going to sleep.
 * Case 2. The new work is spawned after the runtime calls `transition_worker_to_parked`. In this case, the thread that spawned new work will wake up the thread, preventing it from going to sleep. (Or waking it up if it already has gone to sleep.)

In either case, the thread does not actually go to sleep if new work is spawned.
